File: lab-redes/java/tcp/ServidorTCP.java
import java.io.*;
import java.net.*;
import java.time.LocalTime;

public class ServidorTCP {
    public static void main(String[] args) throws IOException {
        int porta = 5000;
        try (ServerSocket servidor = new ServerSocket(porta)) {
            System.out.println("[TCP] Servidor aguardando conexões na porta " + porta + "...");
            try (Socket cliente = servidor.accept();
                 BufferedReader entrada = new BufferedReader(
                         new InputStreamReader(cliente.getInputStream()));
                 PrintWriter saida = new PrintWriter(cliente.getOutputStream(), true)) {

                System.out.println("[TCP] Cliente conectado: " + cliente.getRemoteSocketAddress());
                String mensagem;
                while ((mensagem = entrada.readLine()) != null) {
                    if(mensagem.equalsIgnoreCase("hora")) {
                        String horaAtual = LocalTime.now().toString().substring(0, 8);
                        saida.println("Hora atual: " + horaAtual);
                        System.out.println("[TCP] Recebido: " + horaAtual);
                        continue;
                    }
                    System.out.println("[TCP] Recebido: " + mensagem);
                    if (mensagem.equalsIgnoreCase("sair")) {
                        saida.println("Encerrando conexão. Até mais!");
                        break;
                    }
                    saida.println("Monitor responde: recebi sua mensagem -> \"" + mensagem + "\"");
                }
            }
        }
        System.out.println("[TCP] Servidor encerrado.");
    }
}

File: lab-redes/python/tcp/servidor_tcp.py
import socket

HOST = "0.0.0.0"
PORTA = 5001

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as servidor:
    servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    servidor.bind((HOST, PORTA))
    servidor.listen(1)
    print(f"[TCP] Servidor aguardando conexões na porta {PORTA}...")

    conexao, endereco = servidor.accept()
    with conexao:
        print(f"[TCP] Cliente conectado: {endereco}")
        while True:
            dados = conexao.recv(1024).decode("utf-8").strip()
            if not dados:
                break
            if dados.lower() == "hora":
                from datetime import datetime
                hora_atual = datetime.now().strftime("%H:%M:%S")
                conexao.sendall(f"Hora atual: {hora_atual}\n".encode("utf-8"))
                print(f"[TCP] Enviado: {hora_atual}")
                continue
            print(f"[TCP] Recebido: {dados}")
            if dados.lower() == "sair":
                conexao.sendall("Encerrando conexão. Até mais!\n".encode("utf-8"))
                break
            resposta = f'Monitor responde: recebi sua mensagem -> "{dados}"\n'
            conexao.sendall(resposta.encode("utf-8"))

print("[TCP] Servidor encerrado.")